compiler/rustc_infer/src/errors/mod.rs

```diff
@@ -497,3 +498,350 @@ pub struct MismatchedStaticLifetime<'a> {
+        }
+        diag.span_help(type_param_span, fluent::infer::tid_param_help);
+    }
+}
+
+#[derive(Subdiagnostic)]
+#[help(infer::tid_rel_help)]
+pub struct RelationshipHelp;
+
+#[derive(Diagnostic)]
+#[diag(infer::trait_impl_diff)]
+pub struct TraitImplDiff {
+    #[primary_span]
+    #[label(infer::found)]
+    pub sp: Span,
+    #[label(infer::expected)]
+    pub trait_sp: Span,
+    #[note(infer::expected_found)]
+    pub note: (),
+    #[subdiagnostic]
+    pub param_help: ConsiderBorrowingParamHelp,
+    #[subdiagnostic]
+    // Seems like subdiagnostics are always pushed to the end, so this one
+    // also has to be a subdiagnostic to maintain order.
+    pub rel_help: Option<RelationshipHelp>,
+    pub expected: String,
+    pub found: String,
+}
+
+pub struct DynTraitConstraintSuggestion {
+    pub span: Span,
+    pub ident: Ident,
+}
+
+impl AddToDiagnostic for DynTraitConstraintSuggestion {
+    fn add_to_diagnostic(self, diag: &mut rustc_errors::Diagnostic) {
+        let mut multi_span: MultiSpan = vec![self.span].into();
+        multi_span.push_span_label(self.span, fluent::infer::dtcs_has_lifetime_req_label);
+        multi_span.push_span_label(self.ident.span, fluent::infer::dtcs_introduces_requirement);
+        diag.span_note(multi_span, fluent::infer::dtcs_has_req_note);
+        diag.span_suggestion_verbose(
+            self.span.shrink_to_hi(),
+            fluent::infer::dtcs_suggestion,
+            " + '_",
+            Applicability::MaybeIncorrect,
+        );
+    }
+}
+
+#[derive(Diagnostic)]
+#[diag(infer::but_calling_introduces, code = "E0772")]
+pub struct ButCallingIntroduces {
+    #[label(infer::label1)]
+    pub param_ty_span: Span,
+    #[primary_span]
+    #[label(infer::label2)]
+    pub cause_span: Span,
+
+    pub has_param_name: bool,
+    pub param_name: String,
+    pub has_lifetime: bool,
+    pub lifetime: String,
+    pub assoc_item: Symbol,
+    pub has_impl_path: bool,
+    pub impl_path: String,
+}
+
+pub struct ReqIntroducedLocations {
+    pub span: MultiSpan,
+    pub spans: Vec<Span>,
+    pub fn_decl_span: Span,
+    pub cause_span: Span,
+    pub add_label: bool,
+}
+
+impl AddToDiagnostic for ReqIntroducedLocations {
+    fn add_to_diagnostic(mut self, diag: &mut rustc_errors::Diagnostic) {
+        for sp in self.spans {
+            self.span.push_span_label(sp, fluent::infer::ril_introduced_here);
+        }
+
+        if self.add_label {
+            self.span.push_span_label(self.fn_decl_span, fluent::infer::ril_introduced_by);
+        }
+        self.span.push_span_label(self.cause_span, fluent::infer::ril_because_of);
+        diag.span_note(self.span, fluent::infer::ril_static_introduced_by);
+    }
+}
+
+pub struct MoreTargeted {
+    pub ident: Symbol,
+}
+
+impl AddToDiagnostic for MoreTargeted {
+    fn add_to_diagnostic(self, diag: &mut rustc_errors::Diagnostic) {
+        diag.code(rustc_errors::error_code!(E0772));
```

Comment:
Is it the `.code` here that makes this require a manual impl?

Reply:
It is. This could be solved with two diagnostics which would have a different error code and are the same otherwise, but we only get to know which variant it is only after we create it and call several functions on it, so it's not really feasible.